cdc/sink/mqProducer/kafka.go

```diff
@@ -40,112 +41,248 @@ type kafkaSaramaProducer struct {
+			speed := uint64(0)
+			var speedMB float64
+			if seconds > 0 {
+				qps = count / seconds
+				speed = countSize / seconds
+				speedMB = float64(speed) / float64(1024*1024)
+			}
+
+			lastCount = total
+			lastSize = totalSize
+			lastTime = now
+			log.Info("MQ sink replication status",
+				zap.Uint64("count", count),
+				zap.Uint64("qps", qps), zap.Float64("speed(MB/S)", speedMB))
+		}
 	}
-	return lastIndex, err
 }
 
-func (k *kafkaSaramaProducer) SyncBroadcastMessage(ctx context.Context, key []byte, value []byte) error {
-	wg, cctx := errgroup.WithContext(ctx)
-	for i := int32(0); i < k.partitionNum; i++ {
-		i := i
+func (k *kafkaSaramaProducer) Count() uint64 {
+	return atomic.LoadUint64(&k.count)
+}
+
+func (k *kafkaSaramaProducer) SyncBroadcastMessage(ctx context.Context, key *model.MqMessageKey, value *model.MqMessageDDL) error {
+	wg, _ := errgroup.WithContext(ctx)
+	var valueByte []byte
+	var err error
+	if value != nil {
+		valueByte, err = value.Encode()
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	for partition := 0; partition < int(k.partitionNum); partition++ {
+		partition := partition
 		wg.Go(func() error {
-			var err1, err2 error
 			done := make(chan struct{})
-			_, err1 = k.SendMessage(cctx, key, value, i, func(err error) {
-				err2 = err
-				close(done)
-			})
-			if err1 != nil {
-				return err1
+			var err1 error
+			k.rowPartitionCh[partition] <- kafkaMsg{
+				key:        key,
+				valueByte:  valueByte,
+				checkpoint: true,
+				cb: func(err error) {
+					err1 = err
+					close(done)
+				},
 			}
 			<-done
-			return err2
+			return err1
 		})
 	}
 	return wg.Wait()
 }
 
-func (k *kafkaSaramaProducer) MaxSuccessesIndex() uint64 {
-	maxSentIndex := uint64(0)
-	minSucceededIndex := uint64(math.MaxUint64)
-	for i := 0; i < int(k.partitionNum); i++ {
-		succeedIndex := atomic.LoadUint64(&k.partitionMaxSucceedIndex[i])
-		sentIndex := atomic.LoadUint64(&k.partitionMaxSentIndex[i])
-		if maxSentIndex < sentIndex {
-			maxSentIndex = sentIndex
-		}
-		// if succeedIndex is equal to sentIndex, it means that all of the msgs are sent in this partition,
-		if minSucceededIndex > succeedIndex && succeedIndex != sentIndex {
-			minSucceededIndex = succeedIndex
+func (k *kafkaSaramaProducer) Successes() chan uint64 {
+	return k.successes
+}
+
+func (k *kafkaSaramaProducer) Run(ctx context.Context) error {
+
+	errg, ctx := errgroup.WithContext(ctx)
+	errg.Go(func() error {
+		partitionResolved := make(map[int32]uint64)
+		for {
+			select {
+			case <-ctx.Done():
+				return errors.Trace(ctx.Err())
+			case <-k.closeCh:
+				return nil
+			case msg := <-k.asyncClient.Successes():
+				if msg.Metadata != nil {
+					checkpointTs := msg.Metadata.(uint64)
+					if checkpointTs > partitionResolved[msg.Partition] {
+						partitionResolved[msg.Partition] = checkpointTs
+						// When all the partitions have received at least one resolved ts
+						if len(partitionResolved) == int(k.partitionNum) {
```

Comment:
Seems if no data sent in part of partitions and first `EmitCheckpointEvent` is not fired, the condition will be always false here?

Reply:
Right. This is for preventing a bigger resolved ts is sent back since sarama doesn't guarantee the ordering cross partitions, Should we initialize this map by a more proper value?